algorithmic_efficiency/workloads/wmt/wmt_pytorch/workload.py

```diff
@@ -289,18 +298,26 @@ def _build_input_queue(self,
           batch[key] = tensor[n][RANK]
       yield batch
 
-  def eval_step(self, params, batch):
+  def eval_step(self,
+                params: spec.ParameterContainer,
+                batch: Dict[str, spec.Tensor]) -> Dict[str, spec.Tensor]:
     """Calculate evaluation metrics on a batch."""
     targets = batch['targets']
-    weights = torch.where(targets > 0, 1.0, 0.0)
-    remainder_mask = batch.get('weights')
-    if remainder_mask is not None:
-      weights = torch.logical_and(weights, remainder_mask)
+    weights = batch.get('weights')
     logits, _ = self.model_fn(
         params,
         batch,
         mode=spec.ForwardPassMode.EVAL,
         model_state=None,
         rng=None,
         update_batch_norm=False)
-    return self.compute_summed_metrics(logits, targets, weights)
+    loss = self.compute_weighted_cross_entropy(logits, targets, weights, 0.0)
+    mask = torch.where(targets > 0, 1, 0)
+    if weights is not None:
+      mask = torch.logical_and(weights, mask)
+    acc_sum, weight_sum = self.compute_weighted_accuracy(logits, targets, mask)
+    return {
+        'loss': torch.nansum(loss),
```

Comment:
Ah nice catch, you are completely right! This also affects the change in the OGBG workload (see [here](https://github.com/mlcommons/algorithmic-efficiency/blob/main/algorithmic_efficiency/workloads/ogbg/ogbg_pytorch/workload.py#L200)). The reason why I followed the fix, i.e. padding with nans, is that otherwise we require the user to not use the mean, but to sum the losses and divide by `weights.sum()` (assuming we add the weights in the training `batch`, otherwise we would require the user to know which weights to use). See the discussion in #227. 

Since the instruction on how to average the loss should be workload agnostic, we would have to provide very specific instructions, like saying the users have to set `denominator = batch.get('weights').sum() if 'weights' in batch else len(losses)` or something like this. If the user doesn't follow the instruction there won't be any obvious error; hence, we thought padding with nans is better. But I guess we don't have any other option than to provide explicit instructions on how to average the loss?

Reply:
I agree, I think changing the `loss_fn` API to return a tuple is probably best. We could either go with `(safely averaged loss, per-example losses)` or with `(per-examples losses, correct denominator)`. The advantage of the first version is that many (most?) people probably just want the averaged loss anyway, so they can just get the first output element and don't have to do anything more with it. Also, in the case where the user wants to use the per-example losses, the correct denominator can be derived from this case as well, so the information content of both versions is the same.

If you also think that the first version is best, I will implement it for all workloads in this PR.

Reply:
I went ahead and just implemented this. Let me know if you prefer a different solution.

Reply:
yep I prefer the first one too, thanks for implementing :)

Reply:
+1, implement it in a follow-up PR